File: src/main/java/io/dddrive/ddd/model/base/AggregateSPI.java
package io.dddrive.ddd.model.base;

/**
 * This interface defines the internal callbacks for an Aggregate
 * implementation.
 */
public interface AggregateSPI {

	/**
	 * Get the persistence provider specific aggregate state.
	 * 
	 * @return aggregate state
	 */
	Object getAggregateState();

	/**
	 * Initialise the database records of an Aggregate with some basic fields (id,
	 * tenantId) after creation (internal, technical callback).
	 *
	 * @param aggregateId aggregate id
	 * @param tenantId    tenant id
	 */
	void doInit(Integer aggregateId, Integer tenantId);

	/**
	 * Do some work after create, f.ex. fire events, add transition etc.
	 */
	void doAfterCreate();

	/**
	 * Assign Parts to Aggregate lists after Load
	 */
	void doAssignParts();

	/**
	 * Do some work after load.
	 */
	void doAfterLoad();

	/**
	 * Prepare for storage, f.ex. assign seqNr to parts.
	 */
	void doBeforeStore();

	/**
	 * Calculate the search text and token strings (add via addSearchToken,
	 * addSearchText).
	 */
	void doCalcSearch();

	/**
	 * Do some work after store.
	 */
	void doAfterStore();

}

File: fm-dddrive/src/main/java/dddrive/ddd/model/AggregateSPI.kt
package dddrive.ddd.model

/**
 * This interface defines the internal callbacks for an Aggregate
 * implementation.
 */
interface AggregateSPI {

	/**
	 * Generate new part id.
	 *
	 * @param partClass part class
	 * @return new part id
	 */
	fun <P : Part<*>> nextPartId(partClass: Class<P>): Int

	fun doAfterCreate()

	fun beginLoad()

	fun endLoad()

	fun doAfterLoad()

	fun addPropertyChangeListener(listener: dddrive.property.model.PropertyChangeListener)

	fun removePropertyChangeListener(listener: dddrive.property.model.PropertyChangeListener)

	fun doBeforeStore()

	fun doAfterStore()

	fun fireFieldChange(
		op: String,
		path: String,
		value: Any?,
		oldValue: Any?,
		isInCalc: Boolean,
	)

}
